### PR TITLE
HHH-15190 - Remove support for H2 versions older than 1.4.197

### PR DESCRIPTION
**Pull request by jrenaat**
https://hibernate.atlassian.net/browse/HHH-15190
Signed-off-by: Jan Schatteman <jschatte@redhat.com>

### CODE REVIEW

File: hibernate-community-dialects/src/main/java/org/hibernate/community/dialect/H2LegacyDialect.java [hunk 1]
```diff
@@ -0,0 +1,707 @@
...
+
+		// Prior to 1.4.200 the 'cascade' in 'drop table' was implicit
+		cascadeConstraints = version.isSameOrAfter( 1, 4, 200 );
+		// 1.4.200 introduced changes in current_time and current_timestamp
+		useLocalTime = version.isSameOrAfter( 1, 4, 200 );
+
+		if ( version.isSameOrAfter( 1, 4, 32 ) ) {
+			this.sequenceInformationExtractor = version.isSameOrAfter( 1, 4, 201 )
+					? SequenceInformationExtractorLegacyImpl.INSTANCE
+					: SequenceInformationExtractorH2DatabaseImpl.INSTANCE;
```
Comment (katzyn) [inline hibernate-community-dialects/src/main/java/org/hibernate/community/dialect/H2LegacyDialect.java:152]: Actually there is no such version as 1.4.201. The next release after 1.4.200 is 2.0.202.

Maybe I'm missing something, but why `SequenceInformationExtractorLegacyImpl.INSTANCE` is used for newer versions and `SequenceInformationExtractorH2DatabaseImpl.INSTANCE` is used for older versions in this condition?

File: hibernate-community-dialects/src/main/java/org/hibernate/community/dialect/H2LegacyDialect.java [hunk 2]
```diff
@@ -0,0 +1,707 @@
...
+		return getVersion().isSameOrAfter( 2 );
+	}
+
+	@Override
+	protected String columnType(int sqlTypeCode) {
+		switch ( sqlTypeCode ) {
+			// prior to version 2.0, H2 reported NUMERIC columns as DECIMAL,
+			// which caused problems for schema update tool
+			case NUMERIC:
+				return getVersion().isBefore( 2 ) ? columnType( DECIMAL ) : super.columnType( sqlTypeCode );
```
Comment (katzyn) [inline hibernate-community-dialects/src/main/java/org/hibernate/community/dialect/H2LegacyDialect.java:192]: H2 1.4.197, 1.4.198, 1.4.199, and 1.4.200 return either `Types.DECIMAL` or `Types.NUMERIC` depending on configuration, `Types.DECIMAL` is used by default. But I think presence of this configuration option can be ignored, it is rarely used.

H2 2.0.202 and newer versions return `Types.DECIMAL` for `DECIMAL` / `DEC` and `Types.NUMERIC` for `NUMERIC` columns. But their real type is always `NUMERIC`, there is no specialized implementation for `DECIMAL` / `DEC`.

I don't understand why this condition is required for legacy dialect, it looks like it isn't expected to be used with H2 2.\*.\*.

File: hibernate-community-dialects/src/main/java/org/hibernate/community/dialect/H2LegacyDialect.java [hunk 3]
```diff
@@ -0,0 +1,707 @@
...
+	}
+
+	@Override
+	protected void registerColumnTypes(TypeContributions typeContributions, ServiceRegistry serviceRegistry) {
+		super.registerColumnTypes( typeContributions, serviceRegistry );
+		final DdlTypeRegistry ddlTypeRegistry = typeContributions.getTypeConfiguration().getDdlTypeRegistry();
+
+		if ( getVersion().isBefore( 2 ) ) {
+			ddlTypeRegistry.addDescriptor( new DdlTypeImpl( ARRAY, "array", this ) );
+		}
```
Comment (katzyn) [inline hibernate-community-dialects/src/main/java/org/hibernate/community/dialect/H2LegacyDialect.java:227]: The same problem here, I don't understand why legacy dialect needs a check for H2 2.\*.\*.

File: hibernate-community-dialects/src/main/java/org/hibernate/community/dialect/H2LegacyDialect.java [hunk 4]
```diff
@@ -0,0 +1,707 @@
...
+		else {
+			functionFactory.listagg_groupConcat();
+		}
+	}
+
+	@Override
+	public void augmentPhysicalTableTypes(List<String> tableTypesList) {
+		if ( getVersion().isSameOrAfter( 2 ) ) {
+			tableTypesList.add( "BASE TABLE" );
+		}
```
Comment (katzyn) [inline hibernate-community-dialects/src/main/java/org/hibernate/community/dialect/H2LegacyDialect.java:341]: See previous comment.

File: hibernate-community-dialects/src/main/java/org/hibernate/community/dialect/H2LegacyDialect.java [hunk 5]
```diff
@@ -0,0 +1,707 @@
...
+			String columnTypeName,
+			int jdbcTypeCode,
+			int precision,
+			int scale,
+			JdbcTypeRegistry jdbcTypeRegistry) {
+		// As of H2 2.0 we get a FLOAT type code even though it is a DOUBLE
+		if ( jdbcTypeCode == FLOAT && "DOUBLE PRECISION".equals( columnTypeName ) ) {
+			return jdbcTypeRegistry.getDescriptor( DOUBLE );
+		}
+		return super.resolveSqlTypeDescriptor( columnTypeName, jdbcTypeCode, precision, scale, jdbcTypeRegistry );
```
Comment (katzyn) [inline hibernate-community-dialects/src/main/java/org/hibernate/community/dialect/H2LegacyDialect.java:365]: `REAL` is reported as `Types.REAL`.
`DOUBLE PRECISION` is reported as `Types.DOUBLE`.

`FLOAT` is reported as `Types.DOUBLE` by H2 1.4.196 and older versions.
`FLOAT` is reported as either `Types.REAL` or `Types.DOUBLE` by H2 1.4.197–1.4.200 depending on precision.
`FLOAT` is reported as `Types.FLOAT` by H2 2.0.202 and newer versions, but actual type is either `REAL` or `DOUBLE PRECISION` anyway depending on precision.

File: hibernate-community-dialects/src/main/java/org/hibernate/community/dialect/H2LegacyDialect.java [hunk 6]
```diff
@@ -0,0 +1,707 @@
...
+		switch ( baseTypeName ) {
+			case "CHARACTER VARYING":
+				return VARCHAR;
+		}
+		return super.resolveSqlTypeCode( typeName, baseTypeName, typeConfiguration );
+	}
+
+	@Override
+	public int getMaxVarcharLength() {
+		return 1_048_576;
```
Comment (katzyn) [inline hibernate-community-dialects/src/main/java/org/hibernate/community/dialect/H2LegacyDialect.java:379]: The actual limit for non-LOB data type declarations is
2,147,483,647 characters in H2 1.4.200 and older versions
1,048,576 characters in H2 2.0.202–2.1.210.
1,000,000,000 characters in H2 2.1.212 and newer versions.

LOBs allow longer values.

**Review by katzyn — COMMENTED** [review]
Description of HHH-15190 says that
> There is no clear support policy for H2 from the developers

2.1 series of releases is currently supported by H2 community, the last patch release is 2.1.214.

Support for 2.0 series of releases is currently suspended and unlikely will be resumed unless somebody will be able to prove that 2.1 has significant breaking changes. The last release in this series is 2.0.206 and it has known security issues.

1.4.200 and older versions aren't supported by H2 community in any way.

File: hibernate-community-dialects/src/main/java/org/hibernate/community/dialect/H2LegacyDialect.java (hunk 6 above)
Comment (beikov) [reply hibernate-community-dialects/src/main/java/org/hibernate/community/dialect/H2LegacyDialect.java:379]: PRs are welcome ;)

File: hibernate-community-dialects/src/main/java/org/hibernate/community/dialect/H2LegacyDialect.java (hunk 5 above)
Comment (beikov) [reply hibernate-community-dialects/src/main/java/org/hibernate/community/dialect/H2LegacyDialect.java:365]: If you have ideas to improve the dialect, please create an issue for that and provide a PR

**Comment by beikov** [comment]
Hi @katzyn, the intent is to copy the original main dialect/translator as-is into the hibernate-community-dialects project, named like `...LegacyDialect`/`...LegacySqlAstTranslator`, and then change only the main dialect:

* Introduce a minimum version
* Remove code that checks for versions smaller than the minimum version

I suggested to use the version 1.4.197 as minimum version, because that is currently widely used in some other projects and I didn't want to force everyone to update immediately, especially because the update involves a lot of work for some users. This is just a first step though where the Hibernate team tries to setup a baseline of "supported" versions. We can raise the minimum to 2.1.x for the next minor release again if we see adoption of the new version.

File: hibernate-community-dialects/src/main/java/org/hibernate/community/dialect/H2LegacyDialect.java (hunk 4 above)
Comment (beikov) [reply hibernate-community-dialects/src/main/java/org/hibernate/community/dialect/H2LegacyDialect.java:341]: See https://github.com/hibernate/hibernate-orm/pull/5126#issuecomment-1162714063

File: hibernate-community-dialects/src/main/java/org/hibernate/community/dialect/H2LegacyDialect.java (hunk 2 above)
Comment (beikov) [reply hibernate-community-dialects/src/main/java/org/hibernate/community/dialect/H2LegacyDialect.java:192]: If you have ideas to improve the dialect, please create an issue for that and provide a PR

File: hibernate-community-dialects/src/main/java/org/hibernate/community/dialect/H2LegacyDialect.java (hunk 1 above)
Comment (beikov) [reply hibernate-community-dialects/src/main/java/org/hibernate/community/dialect/H2LegacyDialect.java:152]: If you have ideas to improve the dialect, please create an issue for that and provide a PR

**Review by beikov — APPROVED** [review]

**Review by mamadaliev — APPROVED** [review]

File: hibernate-community-dialects/src/main/java/org/hibernate/community/dialect/H2LegacyDialect.java [hunk 7]
```diff
@@ -0,0 +1,707 @@
...
+			functionFactory.format_formatdatetime();
+		}
+		functionFactory.rownum();
+		if ( getVersion().isSameOrAfter( 1, 4, 200 ) ) {
+			functionFactory.windowFunctions();
+			functionFactory.listagg( null );
+			if ( getVersion().isSameOrAfter( 2 ) ) {
+				functionFactory.inverseDistributionOrderedSetAggregates();
+				functionFactory.hypotheticalOrderedSetAggregates();
+			}
```
Comment (beikov) [inline hibernate-community-dialects/src/main/java/org/hibernate/community/dialect/H2LegacyDialect.java:330]: Comparing this with the current state of `H2Dialect`, it looks like this should rather be the following:

```suggestion
			if ( getVersion().isSameOrAfter( 2 ) ) {
				functionFactory.listagg( null );
				functionFactory.inverseDistributionOrderedSetAggregates();
				functionFactory.hypotheticalOrderedSetAggregates();
			}
			else {
				// Use group_concat until 2.x as listagg was buggy
				functionFactory.listagg_groupConcat();
			}
```

File: hibernate-community-dialects/src/main/java/org/hibernate/community/dialect/H2LegacySqlAstTranslator.java [hunk 8]
```diff
@@ -0,0 +1,213 @@
...
+		return getDialect().getVersion().isSameOrAfter( 1, 4, 197 );
+	}
+
+	@Override
+	protected boolean supportsRowValueConstructorSyntaxInQuantifiedPredicates() {
+		// Just a guess
+		return getDialect().getVersion().isSameOrAfter( 1, 4, 197 );
+	}
+
+	@Override
```
Comment (beikov) [inline hibernate-community-dialects/src/main/java/org/hibernate/community/dialect/H2LegacySqlAstTranslator.java:199]: Looks like there is something missing here from the current state of `H2Dialect`

```suggestion
	@Override
	protected boolean supportsNullPrecedence() {
		// Support for nulls clause in listagg was added in 2.0
		return getClauseStack().getCurrent() != Clause.WITHIN_GROUP || getDialect().getVersion().isSameOrAfter( 2 );
	}

	@Override
```

File: hibernate-core/src/main/java/org/hibernate/dialect/H2Dialect.java [hunk 9]
```diff
@@ -112,20 +113,14 @@ public H2Dialect(DialectResolutionInfo info) {
...
 
 		// https://github.com/h2database/h2database/commit/b2cdf84e0b84eb8a482fa7dccdccc1ab95241440
-		limitHandler = version.isSameOrAfter( 1, 4, 195 )
-				? OffsetFetchLimitHandler.INSTANCE
-				: LimitOffsetLimitHandler.INSTANCE;
-
-		if ( version.isBefore( 1, 2, 139 ) ) {
-			LOG.unsupportedMultiTableBulkHqlJpaql( version.getMajor(), version.getMinor(), version.getMicro() );
-		}
+		limitHandler = OffsetFetchLimitHandler.INSTANCE;
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/dialect/H2Dialect.java:123]: Please remove the field and simply return `OffsetFetchLimitHandler.INSTANCE` from the `getLimitHandler` method.

File: hibernate-core/src/main/java/org/hibernate/dialect/H2Dialect.java [hunk 10]
```diff
@@ -138,16 +133,10 @@ public H2Dialect(DatabaseVersion version) {
...
-					: SequenceInformationExtractorH2DatabaseImpl.INSTANCE;
-			this.querySequenceString = "select * from INFORMATION_SCHEMA.SEQUENCES";
-		}
-		else {
-			this.sequenceInformationExtractor = SequenceInformationExtractorNoOpImpl.INSTANCE;
-			this.querySequenceString = null;
-		}
+		this.sequenceInformationExtractor = version.isSameOrAfter( 1, 4, 201 )
+				? SequenceInformationExtractorLegacyImpl.INSTANCE
+				: SequenceInformationExtractorH2DatabaseImpl.INSTANCE;
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/dialect/H2Dialect.java:138]: Please remove the field and simply return `SequenceInformationExtractorLegacyImpl.INSTANCE` from the `getSequenceInformationExtractor` method.

File: hibernate-core/src/main/java/org/hibernate/dialect/H2Dialect.java [hunk 11]
```diff
@@ -138,16 +133,10 @@ public H2Dialect(DatabaseVersion version) {
...
+		this.querySequenceString = "select * from INFORMATION_SCHEMA.SEQUENCES";
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/dialect/H2Dialect.java:139]: Please remove the field and simply return `"select * from INFORMATION_SCHEMA.SEQUENCES"` from the `getQuerySequencesString` method.

**Review by beikov — CHANGES_REQUESTED** [review]
Seems that some of the original `H2Dialect` and `H2SqlAstTranslator` code was wrongly dropped in the legacy copies. Please fix that. Then we can merge this.

**Review by beikov — CHANGES_REQUESTED** [review]
Seems that some of the original `H2Dialect` and `H2SqlAstTranslator` code was wrongly dropped in the legacy copies. Please fix that. Then we can merge this.

File: hibernate-core/src/main/java/org/hibernate/dialect/H2Dialect.java (hunk 10 above)
Comment (jrenaat) [reply hibernate-core/src/main/java/org/hibernate/dialect/H2Dialect.java:138]: Are you sure about this change? the minimum version will be 1.4.197, i.e. could be less than the 201 that `SequenceInformationExtractorLegacyImpl`requires.

File: hibernate-core/src/main/java/org/hibernate/dialect/H2Dialect.java (hunk 11 above)
Comment (jrenaat) [reply hibernate-core/src/main/java/org/hibernate/dialect/H2Dialect.java:139]: Same comment as above

File: hibernate-core/src/main/java/org/hibernate/dialect/H2Dialect.java (hunk 10 above)
Comment (beikov) [reply hibernate-core/src/main/java/org/hibernate/dialect/H2Dialect.java:133]: Oops, yeah you are right. Looked at the wrong version.

**Pull request merged**